proposals/call-tags/Overview.md

```diff
@@ -4,183 +4,254 @@
+When creating a closure, have it handle all the applicable `ref` representation signatures, but also have it handle the representation signatures for which avoiding the boxing has clear value. (For example, the `funcref` for `(+.)` would handle all the combinations of `i64` and `ref`.)
+For the call tags using non-`ref` representation signatures, have the fall-back box the values into `ref` and then call the given `funcref` with the corresponding `ref`-only representation signature.
+In this way, the caller can *try* to use a specialized representation, and if the callee does not handle that specialization then the fall-back takes care of doing the boxing the caller would have had to do anyways and defer to the uniform representation that everyone is guaranteed to handle.
+This approach also supports separately adding more specialized representations, like `i128`, in dynamically loaded libraries (though, again, canonical call tags cannot be used).
+
+### JavaScript Interop
+
+At present the JS API forces JavaScript functions to be converted to a single typed function signature in order to acquire a corresponding `funcref`.
+This places a burden on tooling, and it underutilizes JavaScript's dynamic typing.
+With call tags, we can simply convert JavaScript functions straight to dynamically typed `funcref`s.
+Similar to how call tags have fall-back handlers, every call tag has a designated behavior for when it is used on one of these JavaScript `funcref`s.
+By default, that behavior is to simply perform the coercions from wasm values to JavaScript values according to the existing JS API, then call the JavaScript function with those values, and then coerce the resulting JavaScript value to wasm values according to the existing JS API.
+That is, the default behavior is the same as if the JavaScript function had been coerced to have the signature of the call tag.
+This can easily be implemented efficiently by generating this coercion code when the call tag is constructed (or first used in this manner).
+For non-canonical call tags, the generating instance can choose to specify a custom coercion (e.g. lossily converting `i64` to `Number` rather than `BigInt`) in a manner to be determined.
+
+### Optimization
+
+If a function has no associated call tags, then it is only directly callable.
+Even if a function does have an associated call tag, if that tag is neither imported nor exported than the optimizer can guarantee that the function can only be successfully called from within the current module instance, and furthermore it can only can be indirectly called from occurrences of `call_with_tag` using one of the associated call tags.
+Both of these help support program analyses and implementation optimizations.
+
+### Abstraction
 
-## Extension: Fall-Back Handlers
+`call_tag.canon` is restricted to providing a canonical call tag for all "concrete" types, such as primitives or imported concrete nominal types (like `$java.lang.String`).
```

Comment:
I'm not clear on how it knows what imported types are abstract vs concrete. Is this something I'm forgetting out of the type imports proposal?

Reply:
Yeah, this abstract vs concrete import concept is just floating around right now. See https://github.com/WebAssembly/gc/pull/241#issuecomment-916148312 and https://github.com/WebAssembly/gc/pull/241#issuecomment-918333086 for more on it.